File: dotnet/src/Envdoctor/Parsers/DockerComposeParser.cs
using System.Text.RegularExpressions;
using Envdoctor.Models;

namespace Envdoctor.Parsers;

/// Parser for docker-compose files.
///
/// Definitions come from `services.<name>.environment:` blocks (both the map
/// and the list form). Bare list entries (`- FOO`) become value-less
/// definitions. `$VAR` / `${VAR}` interpolation anywhere in the file becomes
/// usages.
public sealed class DockerComposeParser : IParser
{
    private static readonly HashSet<string> ComposeBasenames = new()
    {
        "docker-compose.yml",
        "docker-compose.yaml",
        "docker-compose.override.yml",
        "docker-compose.override.yaml",
        "compose.yml",
        "compose.yaml",
    };

    public string Id => "docker-compose";

    public bool MatchPath(string filePath) => ComposeBasenames.Contains(Path.GetFileName(filePath));

    public EnvironmentFile Parse(string content, string filePath)
    {
        var doc = YamlFacade.LoadFirst(content);
        var variables = new List<EnvironmentVariable>();

        if (doc is Dictionary<string, object?> root &&
            root.TryGetValue("services", out var servicesObj) &&
            servicesObj is Dictionary<string, object?> services)
        {
            foreach (var (_, serviceValue) in services)
            {
                object? env = null;
                if (serviceValue is Dictionary<string, object?> svc)
                    svc.TryGetValue("environment", out env);

                foreach (var entry in NormalizeEnvironment(env, content, filePath))
                    variables.Add(entry);
            }
        }

        // `$VAR` / `${VAR}` interpolation → usages.
        var usages = new List<EnvironmentVariable>();
        foreach (var interp in YamlInterp.ScanInterpolations(content))
        {
            var origin = new Origin
            {
                FilePath = filePath,
                Line = interp.Line,
                Kind = OriginKind.Usage,
                Format = OriginFormat.DockerCompose,
            };
            usages.Add(EnvironmentVariable.Create(interp.Name, null, new List<Origin> { origin }));
        }

        return new EnvironmentFile
        {
            FilePath = filePath,
            Format = FileFormat.DockerCompose,
            Variables = EnvironmentVariable.Merge(variables),
            Usages = EnvironmentVariable.Merge(usages),
        };
    }

    /// Flatten a service's `environment:` value into definition variables.
    private static List<EnvironmentVariable> NormalizeEnvironment(object? env, string content, string filePath)
    {
        var variables = new List<EnvironmentVariable>();
        if (env is null)
            return variables;

        if (env is Dictionary<string, object?> map)
        {
            // Map form: KEY: value
            foreach (var (key, rawValue) in map)
            {
                string? value = rawValue is null ? null : YamlFacade.JsString(rawValue);
                var origin = new Origin
                {
                    FilePath = filePath,
                    Line = LineForName(content, key),
                    Kind = value is null ? OriginKind.Reference : OriginKind.Definition,
                    Format = OriginFormat.DockerCompose,
                };
                variables.Add(EnvironmentVariable.Create(key, value, new List<Origin> { origin }));
            }
        }
        else if (env is List<object?> list)
        {
            // List form: - KEY=value | - KEY
            foreach (var item in list)
            {
                if (item is not string s)
                    continue;
                var trimmed = s.Trim();
                if (trimmed.Length == 0)
                    continue;
                var eq = trimmed.IndexOf('=');
                string key;
                string? value;
                if (eq < 0)
                {
                    key = trimmed;
                    value = null;
                }
                else
                {
                    key = trimmed[..eq];
                    value = trimmed[(eq + 1)..];
                }
                var origin = new Origin
                {
                    FilePath = filePath,
                    Line = LineForName(content, key),
                    Kind = value is null ? OriginKind.Reference : OriginKind.Definition,
                    Format = OriginFormat.DockerCompose,
                };
                variables.Add(EnvironmentVariable.Create(key, value, new List<Origin> { origin }));
            }
        }

        return variables;
    }

    /// Best-effort line lookup for a definition name in the raw YAML text.
    private static int? LineForName(string content, string name)
    {
        var escaped = Regex.Escape(name);
        // Supports both map form (`KEY: value`) and list form (`- KEY=value`).
        var re = new Regex($@"^\s*[- ]*[""']?{escaped}[""']?\s*[:=]", RegexOptions.Multiline);
        var match = re.Match(content);
        return match.Success ? YamlInterp.LineForOffset(content, match.Index) : null;
    }
}

File: dart/lib/src/parsers/docker_compose_parser.dart
/// Parser for docker-compose files.
///
/// Definitions come from `services.<name>.environment:` blocks (both the map
/// and the list form). Bare list entries (`- FOO`) become value-less
/// definitions. `$VAR` / `${VAR}` interpolation anywhere in the file becomes
/// usages.
library;

import '../models/environment_file.dart';
import '../models/environment_variable.dart';
import '../models/origin.dart';
import '../utils/paths.dart';
import 'parser.dart';
import 'yaml_facade.dart';
import 'yaml_interp.dart';

class DockerComposeParser implements Parser {
  static const Set<String> composeBasenames = {
    'docker-compose.yml',
    'docker-compose.yaml',
    'docker-compose.override.yml',
    'docker-compose.override.yaml',
    'compose.yml',
    'compose.yaml',
  };

  @override
  String get id => 'docker-compose';

  @override
  bool matchPath(String filePath) => composeBasenames.contains(basename(filePath));

  @override
  EnvironmentFile parse(String content, String filePath) {
    final doc = yamlLoadFirst(content);
    final variables = <EnvironmentVariable>[];

    if (doc is Map<String, Object?>) {
      final services = doc['services'];
      if (services is Map<String, Object?>) {
        for (final serviceValue in services.values) {
          Object? env;
          if (serviceValue is Map<String, Object?>) {
            env = serviceValue['environment'];
          }
          variables.addAll(_normalizeEnvironment(env, content, filePath));
        }
      }
    }

    // `$VAR` / `${VAR}` interpolation → usages.
    final usages = <EnvironmentVariable>[];
    for (final interp in scanInterpolations(content)) {
      final origin = Origin(
        filePath: filePath,
        line: interp.line,
        kind: OriginKind.usage,
        format: OriginFormat.dockerCompose,
      );
      usages.add(EnvironmentVariable.create(interp.name, null, [origin]));
    }

    return EnvironmentFile(
      filePath: filePath,
      format: FileFormat.dockerCompose,
      variables: EnvironmentVariable.merge(variables),
      usages: EnvironmentVariable.merge(usages),
    );
  }

  /// Flatten a service's `environment:` value into definition variables.
  List<EnvironmentVariable> _normalizeEnvironment(
      Object? env, String content, String filePath) {
    final variables = <EnvironmentVariable>[];
    if (env == null) return variables;

    if (env is Map<String, Object?>) {
      // Map form: KEY: value
      for (final entry in env.entries) {
        final key = entry.key;
        final rawValue = entry.value;
        final value = rawValue == null ? null : jsString(rawValue);
        final origin = Origin(
          filePath: filePath,
          line: _lineForName(content, key),
          kind: value == null ? OriginKind.reference : OriginKind.definition,
          format: OriginFormat.dockerCompose,
        );
        variables.add(EnvironmentVariable.create(key, value, [origin]));
      }
    } else if (env is List<Object?>) {
      // List form: - KEY=value | - KEY
      for (final item in env) {
        if (item is! String) continue;
        final trimmed = item.trim();
        if (trimmed.isEmpty) continue;
        final eq = trimmed.indexOf('=');
        String key;
        String? value;
        if (eq < 0) {
          key = trimmed;
          value = null;
        } else {
          key = trimmed.substring(0, eq);
          value = trimmed.substring(eq + 1);
        }
        final origin = Origin(
          filePath: filePath,
          line: _lineForName(content, key),
          kind: value == null ? OriginKind.reference : OriginKind.definition,
          format: OriginFormat.dockerCompose,
        );
        variables.add(EnvironmentVariable.create(key, value, [origin]));
      }
    }

    return variables;
  }

  /// Best-effort line lookup for a definition name in the raw YAML text.
  static int? _lineForName(String content, String name) {
    final escaped = RegExp.escape(name);
    // Supports both map form (`KEY: value`) and list form (`- KEY=value`).
    final re = RegExp('^\\s*[- ]*[""\']?$escaped[""\']?\\s*[:=]', multiLine: true);
    final match = re.firstMatch(content);
    return match == null ? null : lineForOffset(content, match.start);
  }
}
